core/src/main/java/aima/core/search/basic/support/BasicPriorityFrontierQueue.java

```diff
@@ -0,0 +1,220 @@
+package aima.core.search.basic.support;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.TreeSet;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import aima.core.search.api.Node;
+
+/**
+ * An implementation of the Queue interface that wraps an underlying TreeSet
+ * implementation to create a Priority Frontier Queue, which is a Priority 
+ * queue that also, automatically replaces nodes of the same state that 
+ * are already present, but with a lower priority, and does so quickly.
+ * This also has a faster contains function compared to that of the library's
+ * priority queue. All operations are O(log N) in complexity<br> 
+ * 
+ * @author Avinash Agarwal
+ */
+public class BasicPriorityFrontierQueue<A, S> implements Queue<Node<A, S>>, Serializable {
+	
+	private static final long serialVersionUID = 1L;
+	//Red black tree representation for our queue
+	private TreeSet<Node<A, S>> queue;
+	//This maps states with their corresponding nodes that are present in the queue
+	private Map<S, TreeSet<Node<A, S>>> stateNodeMap;
+	
+	public BasicPriorityFrontierQueue() {
+		this(PriorityQueue::new, HashMap::new);
+	}
+	
+	public BasicPriorityFrontierQueue(Supplier<PriorityQueue<Node<A, S>>> underlyingQueueSupplier,
+			Supplier<Map<S, TreeSet<Node<A, S>>>> stateMembershipSupplier) {
+		PriorityQueue<Node<A, S>> temp = underlyingQueueSupplier.get();
+		this.queue = new TreeSet<Node<A, S>>(temp.comparator());
+		queue.addAll(temp);
+		this.stateNodeMap = stateMembershipSupplier.get();
+	}
+
+	//
+	// Queue
+	@Override
+	public boolean add(Node<A, S> node) {
+		
+		//iterating until there is no node with same state and lower priority
+		while(stateNodeMap.containsKey(node.state())) {
+			//find a node having the same state and lower priority (or a node that compares higher)
+			Node<A,S> toDelete = (stateNodeMap.get(node.state())).higher(node);
+			if(toDelete == null) {

```

Comment:
Moved to a separate function.
Also, Directly removing from the NavigableSet returned by tailSet() caused ConcurrentModificationException, therefore i've used .toArray() on it first.